pkg/tools/closectx/context.go

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2021 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package closectx provides helper method to create Close/Unregister context in case of Request/Register failure
+package closectx
+
+import (
+	"context"
+	"time"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/clock"
+	"github.com/networkservicemesh/sdk/pkg/tools/extend"
+)
+
+const closeTimeout = 15 * time.Second
+
+// New creates a new context for Close/Unregister in case of Request/Register failure
+func New(ctx context.Context) (context.Context, context.CancelFunc) {
+	clockTime := clock.FromContext(ctx)
+
+	var timeout time.Duration
+	if deadline, ok := ctx.Deadline(); ok {
+		timeout = clockTime.Until(deadline)
+	}
+	if timeout < closeTimeout {
+		timeout = closeTimeout
+	}
+
+	ctx = extend.WithValuesFromContext(context.Background(), ctx)
```

Comment:
We don't generally know if extending values is correct or incorrect here.

Reply:
This method provides the same context with postponed deadline and disabled cancels from backward. So I suppose that extending values is correct here.

Reply:
If we need to extend values... we can do that easily inline when using this as well... this is orthogonal to the issue of extending the timer.

Reply:
Created 2 helper methods: `Context` and `ContextWithValues`.